Add Potlock contract

CostinCarabas opened the pull request.

CostinCarabas committed:
Add Potlock contract

sasurobert commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,71 @@
+    #[endpoint(acceptPot)]
+    fn accept_pot(&self, potlock_id: PotlockId) {
+        require_potlock_exists(potlock_id);
+        // TODO: Common fund is another contract?
```
no. lets keep it here.

sasurobert commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,71 @@
+        require_potlock_exists(potlock_id);
+
+        //TODO: Common fund is another contract?
+        //TODO: "will return the fee back to the user"
```
yes. return to user.

sasurobert commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,71 @@
+        //TODO: "will return the fee back to the user"
+
+        //TODO: Should we remove the potlock?
+        self.potlocks().clear_entry(proposal_id);
```
yes. remove it.

sasurobert commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,71 @@
+    #[endpoint(acceptApplication)]
+    fn accept_application(&self, project: Projectid) {
+        require_potlock_exists(potlock_id);
+        // TODO: How should we KYC verification in the SC?
```
no. KYC verification is not in the SC. It comes from the outside.

sasurobert commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    #[endpoint(applyForPot)]
+    fn apply_for_pot(&self, project_name: ManagedBuffer, description: ManagedBuffer) {
```
what SC? there is no need here to have a SC.

CostinCarabas committed:
Refactor donations

CostinCarabas committed:
Fix building issues

CostinCarabas committed:
Merge branch 'fix-unit-tests' into add-potlock-contract

CostinCarabas committed:
Framework upgrade potlock

CostinCarabas committed:
Add blackbox tests

CostinCarabas committed:
Added more unit tests

CostinCarabas committed:
Add several more tests

CostinCarabas committed:
Add interactor

CostinCarabas committed:
Adjust interactors functions

CostinCarabas committed:
Add diferent types of actors from different wallets

dorin-iancu commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,110 @@
+}
+
+#[multiversx_sc::module]
+pub trait PotlockStorage {
```
Use camelCase name for storage mappers base key.

dorin-iancu commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,110 @@
+    fn project_donations(
+        &self,
+        project_id: ProjectId,
+    ) -> MapMapper<ManagedAddress, EsdtTokenPayment>;
```
Unless you've got no other choice, avoid MapMapper.

dorin-iancu commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,64 @@
+        project_name: ManagedBuffer,
+        description: ManagedBuffer,
+    ) {
+        let project_id = self.projects().len() + 1;
+        let owner = self.blockchain().get_caller();
+        let project = Project::new(project_id, potlock_id, project_name, description, owner);
+        self.projects().push(&project);
```
I don't think this respects the specs. I don't see where an admin/owner can accept/reject an application. Docs say this:
```
ApplyForPot@potID@projectName@description - anyone can submit application, but an authority will review the submission and evaluate for eligibility and potential impact
```

CostinCarabas replied:
The `new` function for `Project::new` will create a Project with `Status = Inactive`.
An authority would have to call `endpoint(acceptApplication)` to set the project status to Active

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
Merge pull request #84 from multiversx/potlock-framework-upgrade

Framework upgrade potlock

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
Merge remote-tracking branch 'origin/main' into add-potlock-contract

CostinCarabas committed:
Merge pull request #85 from multiversx/potlock-unit-tests

Add blackbox tests

CostinCarabas committed:
Merge branch 'add-potlock-contract' into potlock-system-tests

CostinCarabas committed:
Merge pull request #86 from multiversx/potlock-system-tests

Potlock system tests

CostinCarabas committed:
Framework upgrade 0.51.1

CostinCarabas committed:
Fix tests

CostinCarabas committed:
Fix clippy

CostinCarabas commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,95 @@
+
+    #[only_admin]
+    #[endpoint(removePot)]
+    fn remove_pot(&self, potlock_id: PotlockId) {
```
I will not clear the pot_propose in the accept_pot. Thanks!

CostinCarabas committed:
Fixes after review

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,100 @@
+        for pp in project_percentage {
+            let (project_id, percentage) = pp.into_tuple();
+            let mut output_payments = ManagedVec::new();
+            for (_, donation) in pot_donations.iter() {
```
There may be a few problems with this way of implementing (as Dorin also suggested), mostly concerning gas usage. If a malicious user donates values from 1000 addresses, the potlock will be blocked, with an outOfGas error.
Maybe rethink the architecture of the storages. One option would be instead of a mapMapper, to save under the potlock_id and user keys, a singleValueMapper of PaymentsVec. This allows you to quickly go to a specific user for donations and also, this endpoint would control the gasCost by clearly providing a list of users, besides the potlockId.

psorinionut commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,111 @@
+
+    #[view(getProjects)]
+    #[storage_mapper("projects")]
+    fn projects(&self) -> VecMapper<Project<Self::Api>>;
```
UnorderedSetMapper here as well.

CostinCarabas replied:
Would get a different behavior as explained to potlocks.

CostinCarabas commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,111 @@
+
+    #[view(getPotlocks)]
+    #[storage_mapper("potlocks")]
+    fn potlocks(&self) -> VecMapper<Pot<Self::Api>>;
```
UnorderedSetMapper will reorder the set.
If the set contains Pots at indexes: 1, 2, 3, 4 and Pot #2 is removed, then the set will be reordered to 1, 4, 3.
When we try to retrieve Pot #2, we will get Pot #4 instead of returning an error with "Pot doesn't exist".

CostinCarabas commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,68 @@
+        );
+        let caller = self.blockchain().get_caller();
+
+        let potlock_id = self.potlocks().len() + 1;
```
Removing the second element from a VecMapper that contains 4 elements (1, 2, 3, 4), we would get 1, 3, 4.
Removing the last element from a VecMapper that contains 4 elements (1, 2, 3, 4), we would get 1, 2, 3. By adding another element, we would get 1, 2, 3, 5.
There is no need for the lastId.

CostinCarabas commented on contracts/potlock/src/potlock_setup.rs:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    //// internal functions
+    fn is_valid_potlock_id(&self, potlock_id: PotlockId) -> bool {
```
See my reply there.

CostinCarabas commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,111 @@
+
+    #[view(feePotPayments)]
+    #[storage_mapper("feePotProposer")]
+    fn fee_pot_proposer(&self, potlock_id: PotlockId) -> SingleValueMapper<ManagedAddress>;
```
Moved to Potlock struct.

CostinCarabas commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,68 @@
+        project_name: ManagedBuffer,
+        description: ManagedBuffer,
+    ) {
+        let project_id = self.projects().len() + 1;
```
See my reply from potlocks.

CostinCarabas committed:
Fixes after review

CostinCarabas committed:
Fix tests

CostinCarabas committed:
Refuse multiple donations with different tokens

CostinCarabas committed:
Add tests for multiple donations with different tokens

CostinCarabas committed:
Fix donate_to_project

Add amount of current payment to previous payment

CostinCarabas commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,68 @@
+        let caller = self.blockchain().get_caller();
+        self.require_potlock_exists(potlock_id);
+        self.require_potlock_is_active(potlock_id);
+        self.pot_donations(potlock_id).insert(caller, payment);
```
Accepted multiple payments but only with the same token_id.

psorinionut replied:
This function is still the same. The payment token is not checked. And in case you only ask for a specific payment token, you would still need a check, to see if there is already some donation from the caller. If it isn't, then you add it, if there's already something, you accumulate it. But this works only if the accepted token_id is always the same and not changeable.

psorinionut replied:
Now I see you did this update for the donateToProject endpoint, but it is needed here also.

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
Framework upgrade 0.52.1

CostinCarabas committed:
potlock: distributePotToProjects: percentage req

Modify require_correct_percentages from distributePotToProjects
endpoint. The total may be lower than 100%.
This enables the caller to distribute part of the pot.

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+    #[endpoint(changeFeeForPots)]
+    fn change_fee_for_pots(&self, token_identifier: TokenIdentifier, fee: BigUint) {
+        require!(
+            token_identifier.is_valid_esdt_identifier(),
```
This check is duplicated.

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+            "Invalid token provided"
+        );
+        require!(
+            token_identifier.is_valid_esdt_identifier() && fee.ge(&BigUint::zero()),
```
Also, here the fee shouldn't be strictly greater than 0? Right now it checks if it is greater or equal. For better clarity, I would keep the checks in separate lines. So one for valid esdt and the other one for >0 check.

dorin-iancu replied:
Btw, simply use the `>` or `>=` operator instead, and just use `0` instead of `BigUint::zero()`, way easier to read that way.

And yeah, agreed with Sorin, keep them in different requires.

psorinionut commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,159 @@
+
+    #[view(feeAmountAcceptPots)]
+    #[storage_mapper("feeAmountAcceptedPots")]
+    fn fee_amount_accepted_pots(&self) -> SingleValueMapper<BigUint>;
```
What's the usage of this storage? You only increment it in the acceptPot, so just to keep track of the total pot amounts over time? Do you need this in the SC stored?

CostinCarabas replied:
Removed.

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,78 @@
+    fn add_pot(&self, name: ManagedBuffer, description: ManagedBuffer) {
+        let payment_for_adding_pot = self.call_value().single_esdt();
+        require!(
+            self.fee_token_identifier().get() == payment_for_adding_pot.token_identifier,
```
Only one payment token for all pots? And an ESDT? So I guess it will be WEGLD?

CostinCarabas replied:
This is the token paid by the pot initiator (from the specs: _AddPot@potDescription - it sets the description and the name of the pot, the call requires a FEE in a defined tokenID._)

The fee token can be set to any token, probably WEGLD.
Another contract can be deployed for xExchange and the fee will probably be MEX.

psorinionut commented on contracts/potlock/interact-rs/src/proxy.rs:
```diff
@@ -0,0 +1,399 @@
+where
+    Api: ManagedTypeApi,
+{
+    pub potlock_id: usize,
```
Why do you save the id of the Pot inside the struct? As it is actually given by the index of its position in the VecMapper.

psorinionut commented on contracts/potlock/interact-rs/src/proxy.rs:
```diff
@@ -0,0 +1,399 @@
+where
+    Api: ManagedTypeApi,
+{
+    pub project_id: usize,
```
Same as with the potlock_id inside the Pot struct.

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,78 @@
+
+        let mut donation_mapper = self.project_donations(project_id);
+        if donation_mapper.contains_key(&caller) {
+            let opt_payment = donation_mapper.remove(&caller);
```
Instead of removing (which updates the storage and is a more expensive operation), I would just read the value (less expensive), as you overwrite it anyway.

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,78 @@
+        project_name: ManagedBuffer,
+        description: ManagedBuffer,
+    ) {
+        let project_id = self.projects().len() + 1;
```
As previously suggested, I don't think there's a need for saving the id inside the struct. And if you remove it, you can remove the len + 1 computation, and just return the usize from the last line
`self.projects().push(&project)`

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+{
+    #[only_admin]
+    #[endpoint(changeFeeForPots)]
+    fn change_fee_for_pots(&self, token_identifier: TokenIdentifier, fee: BigUint) {
```
After a second review, I think we should not change the token_identifier more than once. So once it is set, you cannot update it any longer. So you can use set_if_empty() here, or check that the storage is empty.

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+    #[only_admin]
+    #[endpoint(acceptPot)]
+    fn accept_pot(&self, potlock_id: PotlockId) {
+        self.require_potlock_exists(potlock_id);
```
I would add an extra check here, that the potlock is inactive. So there is no chance for a double call of this endpoint.

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+    #[only_admin]
+    #[endpoint(removePot)]
+    fn remove_pot(&self, potlock_id: PotlockId) {
+        self.require_potlock_exists(potlock_id);
```
I would add an extra check here also, that the potlock is inactive.

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+
+    #[only_admin]
+    #[endpoint(acceptApplication)]
+    fn accept_application(&self, project_id: ProjectId) {
```
Maybe also add another endpoint to reject and remove project applications? Like with the pot?

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+            self.send().direct_multi(&project_owner, &output_payments);
+        }
+
+        self.pot_donations(potlock_id).clear();
```
Also, after the pot is distributed to the projects, it still remains active? So once accepted, the pot remains active indefinitely?
If not, maybe some extra storage clearing should be done for the potlocks mapper?

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+        let pot_donations = self.pot_donations(potlock_id);
+
+        for pp in project_percentages {
+            let (project_id, percentage) = pp.into_tuple();
```
Maybe add an extra check here, if the project_id indeed applied for the given pot? To not give the a part of the share to outside projects.

dorin-iancu committed:
Merge branch 'main' into add-potlock-contract

dorin-iancu commented on contracts/potlock/src/potlock.rs:
```diff
@@ -0,0 +1,28 @@
+pub mod potlock_interactions;
+pub mod potlock_storage;
+
+/// An empty contract. To be used as a template when starting a new contract from scratch.
```
Don't mark as resolved if you don't fix it :nutu:

dorin-iancu commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,111 @@
+    #[endpoint(rejectDonation)]
+    fn reject_donation(&self, potlock_id: PotlockId, user: ManagedAddress) {
+        self.require_potlock_exists(potlock_id);
+        let opt_fee_pot_payments = self.pot_donations(potlock_id).get(&user);
```
Why is it called `payments` if it's a single one?

dorin-iancu commented on contracts/potlock/src/potlock_storage.rs:
```diff
@@ -0,0 +1,159 @@
+
+#[multiversx_sc::module]
+pub trait PotlockStorage {
+    fn is_valid_potlock_id(&self, potlock_id: PotlockId) -> bool {
```
Maybe move those "require" functions to a different file. `PotlockStorage` implies it only contains Storage, but there's a whole lot of other stuff here.

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
potlock: Framework upgrade

CostinCarabas committed:
Fix README

CostinCarabas committed:
Fixes after review

CostinCarabas committed:
Fixes after review

CostinCarabas committed:
Fixes after review

CostinCarabas committed:
Fix tests

CostinCarabas committed:
Fix compiling issues

CostinCarabas committed:
potlock: more unit tests

vladiouz committed:
Add interactor tests

vladiouz committed:
Add interactor tests

vladiouz committed:
Solve warnings

vladiouz committed:
Execute interactor tests only on demand

vladiouz committed:
Cleanse tests

vladiouz committed:
Cleanse tests

vladiouz committed:
Merge pull request #105 from multiversx/add-potlock-contract-interactor-tests

Add potlock contract interactor tests

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
Framework upgrade 0.53.2

CostinCarabas committed:
potlock: clippy fix

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,106 @@
+            donation_mapper.insert(caller, payment);
+        }
+
+        // match donation_mapper.get(&caller) {
```
Delete comment.

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,106 @@
+        self.require_potlock_exists(potlock_id);
+        self.require_potlock_is_active(potlock_id);
+
+        let payment = self.call_value().single_esdt();
```
Do you need to accept NFT payments? If not, I would change this to single_fungible_esdt.

psorinionut commented on contracts/potlock/src/potlock_interactions.rs:
```diff
@@ -0,0 +1,106 @@
+    #[payable("*")]
+    #[endpoint(addPot)]
+    fn add_pot(&self, name: ManagedBuffer, description: ManagedBuffer) {
+        let payment_for_adding_pot = self.call_value().single_esdt();
```
Not mandatory, but you could add these 2 storages under a single one, that keeps a EsdtTokenPayment struct. And you can check here if the payments are equal.

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,132 @@
+
+    #[only_admin]
+    #[endpoint(distributePotToProjects)]
+    fn distribute_pot_to_projects(
```
A few more questions here. Not mandatory but are good to be discussed.
- Right now, there is no way to be sure all the projects that applied for this pot were covered in the provided array, is it ok for the owner to be able to give the amounts arbitrary and to not all projects? 
- Also, there is no check that the project applied to this specific pot, maybe add a check after you read each project from storage, that the project is active for this pot? If not, why even bother to approve a project for a pot?
- And finally, how about a limit to the number of projects that can apply for a pot? Either here for the number of elements for project_percentages, or for when subscribing a project to a pot.

psorinionut commented on contracts/potlock/src/potlock_requirements.rs:
```diff
@@ -0,0 +1,69 @@
+            let (_, perc) = pp.into_tuple();
+            total_perc += perc;
+        }
+        require!(
```
I would still enforce clear equality here.
total_perc == MAX_PERCENTAGE

CostinCarabas replied:
We'll keep it like this for now in case we want to distribute part of the pot.

CostinCarabas committed:
Fixes after review

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
framework upgrade

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
potlock: interactor fixes

CostinCarabas committed:
potlock: remove commented code

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract

psorinionut commented on contracts/potlock/src/potlock_admin_interactions.rs:
```diff
@@ -0,0 +1,132 @@
+            let project = all_projects.get(project_id);
+
+            // The project must previously apply to this Pot
+            if project.potlock_id != potlock_id {
```
Maybe also add a check here that the project is active.

CostinCarabas committed:
Merge branch 'main' into add-potlock-contract
